File: heap_PriorityQueue/cppRealizations/kClosestPointstoOrigin.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;


class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        priority_queue<vector<int>> minHeap;
        for(int i = 0; i < points.size(); i++){
            int dist = points[i][0] * points[i][0] + points[i][1] * points[i][1];
            minHeap.push({-dist, points[i][0], points[i][1]});
        }

        vector<vector<int>> result;
        for(int i = 0; i < k; i++){
            result.push_back({minHeap.top()[1], minHeap.top()[2]});
            minHeap.pop();
        }
        return result;
    }
};

ostream& operator<<(ostream& out, const vector<int>& str){
    out << "[";
    for(size_t i = 0; i < str.size(); i++){
        if(i < str.size() - 1){
            out << str[i] << ",";
        }else{
            out << str[i];
        }
    }
    out << "]";
    return out;
}

ostream& operator<<(ostream& out, const vector<vector<int>>& str){
    out << "[";
    for(size_t i = 0; i < str.size(); i++){
        if(i < str.size() - 1){
            out << str[i] << ",";
        }else{
            out << str[i];
        }
    }
    out << "]";
    return out;
}

int main(){
    vector<vector<int>> points = {{1, 3}, {-2, 2}};
    int k = 1;
    cout << "Input:" << endl;
    cout << "points = " << points << endl;
    vector<vector<int>> result = Solution().kClosest(points, k);
    cout << "kClosest: " << result << endl;
    return 0;
}

File: heap_PriorityQueue/kClosestPointstoOrigin.py
import heapq


class Solution:
    def kClosestMySolution(self, points: list[list[int]], k: int) -> list[list[int]]:
        hashMap = {}
        dists = []
        heapq.heapify(dists)

        for point in points:
            dist = point[0] ** 2 + point[1] ** 2
            if dist not in hashMap:
                hashMap[dist] = [point]
            else:
                hashMap[dist].append(point)
            heapq.heappush(dists, -dist)

            if len(dists) > k:
                dist = -heapq.heappop(dists)
                if len(hashMap[dist]) > 0:
                    hashMap[dist].pop()
        
        result = []
        while k > 0:
            dist = -heapq.heappop(dists)
            if len(hashMap[dist]) > 0:
                result.append(hashMap[dist].pop())
                k -= 1

        return result

    def kClosest(self, points: list[list[int]], k: int) -> list[list[int]]:
        minHeap = []
        for x, y in points:
            dist = (x**2) + (y**2)
            minHeap.append([dist, x, y])

        heapq.heapify(minHeap)
        result = []
        while k > 0:
            dist, x, y = heapq.heappop(minHeap)
            result.append([x, y])
            k -= 1

        return result


if __name__ == "__main__":
    s = Solution()
    print("My Solution:")
    print(s.kClosestMySolution([[1, 3], [-2, 2]], 1))
    print("NeetCode Solution:")
    print(s.kClosest([[1, 3], [-2, 2]], 1))
